### PR TITLE
Check permissions of queries via `EXPLAIN EXECUTE <stmt>;`

### LINKED ISSUE #561: Does this library reflect permissions?
**Is your feature request related to a problem? Please describe.**

Hi, I've tried searching the repository for `grant`, `grants`, `rbac`, `permissions` to no avail; I was wondering if `pgtyped` understands role-based access control; for example:

```sql
create extension if not exists citext;

create table user_emails (
  id int primary key generated always as identity,
  address citext not null unique,
  receives_notifications boolean not,
  created_at timestamptz not null default now(),
  updated_at timestamptz not null default now()
);

revoke all on user_emails from some_role;
grant
  select,
  insert (address, receives_notifications),
  update (receives_notifications),
  delete
on user_emails to some_role;
```

In this setup, we can know that `insert into user_emails (address) values ('test@example.com')` is fine; but `insert into user_emails (id, address) values (1, 'test@example.com')` (when running with `some_role`) is not allowed.

Can `pgtyped` pick up on this? If so, is there some special setup I need? If not, how hard would it be for me to add this, either to core or via an external mechanism?

**Describe the solution you'd like**

I'd like to know that this is valid:

```ts
const query = sql`INSERT INTO user_emails (address) VALUES $$users(address) RETURNING id`;
```

And this should raise a TypeScript error (since insert to `id` column is not granted):

```ts
const query = sql`INSERT INTO user_emails (id, address) VALUES $$users(id, address) RETURNING id`;
```

**Additional context**

I've hardly used `pgtyped` before and have barely scraped the surface of the docs, but I am the author of [PostGraphile](https://postgraphile.org/) (extremely efficient GraphQL API backed primarily by GraphQL, written in TypeScript), [`pg-introspection`](https://star.graphile.org/pg-introspection) (introspection utility for PostgreSQL databases, with TypeScript types built using the PostgreSQL system catalog documentation), [`pg-sql2`](https://star.graphile.org/pg-sql2) (incredibly fast SQL tagged template library for PostgreSQL syntax with minimal allocations/GC overhead) and various other things so I'm no slouch on the Postgres front, and not afraid to get my hands dirty in TypeScript. That said, this is just an information gathering exercise right now.

I searched the codebase for `pg_class` and `information_schema` and on finding no results got really curious and dug a bit deeper... From what I can tell from a quick scan of the code, you seem to be using the raw PostgreSQL protocol to prepare queries and then using the descriptions that PostgreSQL gives you back to determine the parameter/return types? This is a cunning way of side-stepping the need for introspection/a parser/etc! I've not yet dabbled with the wire protocol but it has been on my TODO list for a while now. That said, I don't think the wire protocol contains sufficient detail for this (I assume it contains similar details to the [`pg_prepared_statements` view](https://www.postgresql.org/docs/current/view-pg-prepared-statements.html)) - do you agree? If so, I think we might have a few options available short of adding a parser and introspection library... Would love to discuss further if you're interested.

### PR DESCRIPTION
**Pull request by benjie**
- Fixes #561

This PR adds an additional check before generating the types of a query, it runs something like `EXPLAIN EXECUTE prepared_statement_name (null, null, null, ...);` which doesn't actually run the query (that would be `EXPLAIN ANALYZE EXECUTE prepa...`) but does make the DB attempt to form a query plan for it, which will force some rudimentary query checks including, critically, the assertion of role-based access control.

This PR is not merge-able for (at least) the following reasons:

1. There are negative assertions that I don't understand how to make with your test suite (I want to assert that a given SQL query fails to generate types)
2. This is _currently_ a breaking change for people who are using `NOINHERIT` in their role configuration, since it only checks that the user in the connection string is capable of making the query, not the other roles it might become via `SET ROLE` (solution: opt-in via configuration setting? I've added a dummy variable for this.)
3. Undocumented

Also, potentially the logic of `explainQuery` should be merged into `getTypeData`? That's where I started, but the changes got too large so I figured it would be easier to review separately, plus easier to opt in/out of that way also.

I've done what I can to add to the tests, including creating a database role `pgtyped_test` to use instead of the `postgres` superuser role, making sure all previous tests still pass, adding a passing test about a new constraint `insert` query, and adding what should be negative assertions which I can see are picked up by the system:

```
Error processing src/user_emails/assert_fail.ts: permission denied for table user_emails                                                                                                                                                                          
undefined       
```

but I don't know how to check that or mark it as expected.

Note: for this to work for me I had to change the `test` script in packages/example/package.json` to:

```diff
-    "test": "docker-compose run build && docker-compose run test && docker-compose run test-cjs",
+    "test": "docker compose down && docker compose run build && docker compose run test && docker compose run test-cjs",
```

for two reasons: a) I needed the DB to shut down so it would re-initialize with the `schema.sql` script, and b) I don't have `docker-compose`, only `docker compose`.

Please feel very free to take over this pull request if you wish. Alternatively, if you're interested in supporting this feature, please let me know how I might best fix the issues outlined above.

### CODE REVIEW

File: packages/query/src/actions.ts [hunk 1]
```diff
@@ -250,6 +257,89 @@ export async function getTypeData(
...
+     *   severity: 'ERROR',
+     *   message: 'permission denied for table user_emails'
+     * }
+     * ```
+     */
+    console.error('Error occurred whilst testing query:', e);
+    return {
+      errorCode: '_ERROR',
+      message: e.message,
+    };
```
Comment (benjie) [inline packages/query/src/actions.ts:326]: This does not feel right, but it makes the error handled by `insert into user_emails(created_at)` be handled in the same way as `insert into user_emails(id)`, thereby making it so I can add the functioning tests. The former being an RBAC permission which is thrown at `EXPLAIN EXECUTE` time, whereas the latter is a generated column error which is thrown (I think) at parse time.

**Comment by benjie** [comment]
Another potential issue in this PR: `STABLE` functions can be called at planning time, so if users have `STABLE` functions that they call in their SQL queries and those functions `RAISE EXCEPTION` then the query may be invalidated due to that exception; for example:

```sql
create function current_user_id() returns int as $$
declare
  id int;
begin
  id := nullif(current_setting('jwt.claims.user_id', true), '');
  if id is null then
    raise exception 'FORBIDDEN'; -- WORKAROUND: just return `null`!
  end if;
  return id;
end;
$$ language plpgsql stable;
```

File: packages/query/src/actions.ts [hunk 2]
```diff
@@ -43,6 +43,7 @@ export async function startup(
       user: options.user,
       database: options.dbName,
       client_encoding: "'utf-8'",
+      application_name: 'pgtyped',
```
Comment (benjie) [inline packages/query/src/actions.ts:46]: > `application_name` (string)
>  The `application_name` can be any string of less than `NAMEDATALEN` characters (64 characters in a standard build). **It is typically set by an application upon connection to the server. The name will be displayed in the `pg_stat_activity` view and included in CSV log entries.** It can also be included in regular log entries via the log_line_prefix parameter. Only printable ASCII characters may be used in the application_name value. Other characters are replaced with C-style hexadecimal escapes.
> -- https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-APPLICATION-NAME

It's good practice to set this; but it also give people a way to `return null` rather than `raise exception` from their `STABLE` functions if it's `pgtyped` running.
